File: baekjoon/python/2178/2178.py
from collections import deque
# import sys

# sys.stdin = open("2178_input.txt", "r")

# 최단거리 bfs
# 시작위치, 도착위치 포함 2번째 칸은 거리 2라는 뜻
n, m = map(int, input().split())

load = []

for i in range(n):
    load.append(list(map(int, input())))

# 출발위치 - 0,0 이고, 목적위치 n-1,m-1인 것을 기억
# 벽 0, 길 1


def bfs(x, y):
    q = deque()
    q.append((x, y))
    # 이동범위
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            # 공간 밖이라면
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            # 길이 아니라면
            if load[nx][ny] == 0:
                continue
            # 길이라면
            if load[nx][ny] == 1:
                # 현재 경로 + 1 추가한 거리
                q.append((nx, ny))
                load[nx][ny] = load[x][y] + 1

    print(load[n-1][m-1])


bfs(0, 0)

File: baekjoon/cpp/2178/2178.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int n, m;

vector<int> graph[101];
bool visited[101][101] = {
    false,
};

void bfs(int sx, int sy)
{
    queue<pair<int, int>> q;
    q.push({sx, sy});
    visited[sx][sy] = true;

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};
    int nx, ny;
    while (!q.empty())
    {
        int x, y;
        x = q.front().first;
        y = q.front().second;
        q.pop();
        for (int i = 0; i < 4; i++)
        {
            nx = x + dx[i];
            ny = y + dy[i];
            if (0 <= nx && nx < n && 0 <= ny && ny < m && !visited[nx][ny])
            {
                if (graph[nx][ny])
                {
                    visited[nx][ny] = true;
                    graph[nx][ny] = graph[x][y] + 1;
                    q.push({nx, ny});
                }
            }
        }
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    freopen("2178_input.txt", "r", stdin);

    cin >> n >> m;
    for (int i = 0; i < n; i++)
    {
        string tmp;
        cin >> tmp;
        for (const auto &elem : tmp)
        {
            graph[i].push_back(elem - '0');
        }
    }
    // for (int i = 0; i < n; i++)
    // {
    //     for (const auto &elem : graph[i])
    //     {
    //         cout << elem << " ";
    //     }
    //     cout << endl;
    // }

    bfs(0, 0);

    // for (int i = 0; i < n; i++)
    // {
    //     for (const auto &elem : graph[i])
    //     {
    //         cout << elem << " ";
    //     }
    //     cout << endl;
    // }
    cout << graph[n - 1][m - 1] << endl;
    return 0;
}
